py/builtinimport.c

```diff
@@ -438,31 +439,41 @@ STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name,
-            mp_store_attr(module_obj, MP_QSTR___path__, mp_obj_new_str(vstr_str(path), vstr_len(path)));
-            size_t orig_path_len = path->len;
-            vstr_add_str(path, PATH_SEP_CHAR "__init__.py");
-            if (stat_file_py_or_mpy(path) == MP_IMPORT_STAT_FILE) {
+        nlr_buf_t nlr;
+        if (nlr_push(&nlr) == 0) {
+            if (stat == MP_IMPORT_STAT_DIR) {
+                // Directory -- execute "path/__init__.py".
+                DEBUG_printf("%.*s is dir\n", (int)vstr_len(path), vstr_str(path));
+                // Store the __path__ attribute onto this module.
+                // https://docs.python.org/3/reference/import.html
+                // "Specifically, any module that contains a __path__ attribute is considered a package."
+                mp_store_attr(module_obj, MP_QSTR___path__, mp_obj_new_str(vstr_str(path), vstr_len(path)));
+                size_t orig_path_len = path->len;
+                vstr_add_str(path, PATH_SEP_CHAR "__init__.py");
+                if (stat_file_py_or_mpy(path) == MP_IMPORT_STAT_FILE) {
+                    do_load(MP_OBJ_TO_PTR(module_obj), path);
+                } else {
+                    // No-op. Nothing to load.
+                    // mp_warning("%s is imported as namespace package", vstr_str(&path));
+                }
+                // Remove /__init__.py suffix.
+                path->len = orig_path_len;
+            } else { // MP_IMPORT_STAT_FILE
+                // File -- execute "path.(m)py".
                 do_load(MP_OBJ_TO_PTR(module_obj), path);
-            } else {
-                // No-op. Nothing to load.
-                // mp_warning("%s is imported as namespace package", vstr_str(&path));
+                // Note: This should be the last component in the import path.  If
+                // there are remaining components then it's an ImportError
+                // because the current path(the module that was just loaded) is
+                // not a package.  This will be caught on the next iteration
+                // because the file will not exist.
             }
-            // Remove /__init__.py suffix.
-            path->len = orig_path_len;
-        } else { // MP_IMPORT_STAT_FILE
-            // File -- execute "path.(m)py".
-            do_load(MP_OBJ_TO_PTR(module_obj), path);
-            // Note: This should be the last component in the import path.  If
-            // there are remaining components then it's an ImportError
-            // because the current path(the module that was just loaded) is
-            // not a package.  This will be caught on the next iteration
-            // because the file will not exist.
+            nlr_pop();
+        } else {
+            // An exception happened while running the module code above.
+            // Remove the partially-loaded module from sys.modules so future
+            // imports will attempt to run the code again.
+            mp_obj_module_unregister(full_mod_name);
```

Comment:
Not creating/calling a new function like @stinos suggested saves about 50 bytes on x86_64 Linux.

```suggestion
            mp_obj_dict_delete(MP_OBJ_FROM_PTR(&MP_STATE_VM(mp_loaded_modules_dict)), MP_OBJ_NEW_QSTR(full_mod_name));
```